File: bubble_sort.py
# Python3 Optimized implementation 
# of Bubble sort 
filename = 'data1000.txt'  
# An optimized version of Bubble Sort 
def bubbleSort(arr): 
    n = len(arr) 
   
    # Traverse through all array elements 
    for i in range(n): 
        swapped = False
  
        # Last i elements are already 
        #  in place 
        for j in range(0, n-i-1): 
   
            # traverse the array from 0 to 
            # n-i-1. Swap if the element  
            # found is greater than the 
            # next element 
            if arr[j] > arr[j+1] : 
                arr[j], arr[j+1] = arr[j+1], arr[j] 
                swapped = True
  
        # IF no two elements were swapped 
        # by inner loop, then break 
        if swapped == False: 
            break
           
# Driver code to test above 
import time
start_time = time.time()

f = open(filename,'r')
data = f.read()
arr = [int(i) for i in data.split()]
   
bubbleSort(arr) 
print("Execution time = %s seconds" % (time.time() - start_time))

File: bubble_sort.c
#include <math.h> 
#include <stdio.h> 
#include <stdlib.h>
#include <time.h>
#define NUMEL 1000
char* filename = "data1000.txt";

void swap(int *xp, int *yp) 
{ 
    int temp = *xp; 
    *xp = *yp; 
    *yp = temp; 
} 
  
// A function to implement bubble sort 
void bubbleSort(int arr[], int n) 
{ 
   int i, j; 
   for (i = 0; i < n-1; i++)       
  
       // Last i elements are already in place    
       for (j = 0; j < n-i-1; j++)  
           if (arr[j] > arr[j+1]) 
              swap(&arr[j], &arr[j+1]); 
}  
// A utility function to print an array of size n 
void printArray(int arr[], int n) 
{ 
    int i; 
    for (i = 0; i < n; i++) 
        printf("%d ", arr[i]); 
    printf("\n"); 
} 

void readfile(int A[], int size)
{
	FILE *fp = fopen(filename,"r");
	if (fp == NULL)
		return;
	for (int i = 0; i < size; i++)
	{
		fscanf(fp, "%d", &A[i]);
	}
	fclose(fp);
}
 
/* Driver program to test above functions */
int main() 
{ 
    clock_t start, end;
    start = clock();
    int *arr = malloc(sizeof(int)*NUMEL);  
    int arr_size = NUMEL; 
    readfile(arr, NUMEL);
  
    //printf("Given array is \n"); 
    //printArray(arr, arr_size); 
  
    bubbleSort(arr, NUMEL); 
  
    //printf("\nSorted array is \n"); 
    //printArray(arr, arr_size); 
    free(arr);
    end = clock();
    printf("Execution time = %f seconds\n", ((double) (end - start)) / CLOCKS_PER_SEC);
    return 0; 
} 
